src/Http/Http/src/IFileBufferingStreamFactory.cs

```diff
@@ -0,0 +1,15 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.IO;
+using Microsoft.AspNetCore.WebUtilities;
+
+namespace Microsoft.AspNetCore.Http
+{
+    public interface IFileBufferingStreamFactory
+    {
+        FileBufferingReadStream CreateReadStream(Stream inner, int bufferThreshold, long? bufferLimit = null);
```

Comment:
Why can't this return Stream?

Reply:
The consumers are actually expecting a Stream, so it can return it. The only reason was because the consumers of the writer need the  `FileBufferingWriteStream` explicitly so I kept this one same way. Moving this to return `Stream` sounds better.

https://github.com/aspnet/AspNetCore/blob/6152fc37b7ab4d61ac17f4aa1c5109fd5ee8b2e8/src/Mvc/Mvc.Formatters.Xml/src/XmlDataContractSerializerOutputFormatter.cs#L291

Reply:
> The only reason was because the consumers of the writer need the FileBufferingWriteStream explicitly

I expect we could remove that requirement. DrainBufferAsync looks like a fancy CopyToAsync.

Reply:
> I expect we could remove that requirement. DrainBufferAsync looks like a fancy CopyToAsync.

By default `CopyTo `throws `NotSupportedException` when a stream cannot be read, like in this case. Could it be confusing if we allow to copy this stream but reading it is not allowed? I don't know, however, if this behavior is usually honored.

Reply:
You're right, that came up in the design so we added DrainBufferAsync. I'd rather not have the concrete type exposed here. Maybe we can abstract it.
```
public interface IBufferedWriteStream
{
  Stream Buffer { get; }
  Task DrainBufferAsync(Stream destination);
}
```

Reply:
If I got it right, `FileBufferingWriteStream` would be implementing `IBufferedWriteStream` and returned from the factory instead of the concrete type.

Regarding of `Buffer` property, and also that we need the actual `Stream` in the consumers, what is the purpose of the property and how are we getting the `Stream`?

I have the feeling that property could be the answer since `IBufferedWriteStream` doesn't know it's a `Stream` (cannot extend the class), so TL;DR, could be the implementation this one and the way of getting it instead of casting on the consumers?

`Stream Buffer => this;`

Reply:
> `Stream Buffer => this;`

Exactly